Add STORAGE_BACKEND support in configuration.py

gervaso opened the pull request:
This settings enable the support for STORAGE_BACKEND and STORAGE_CONFIG in the configuration.py file.

<!--
#########################################################################

Thank you for sharing your work and for opening a PR.

(!) IMPORTANT (!):
First make sure that you point your PR to the `develop` branch!

Now please read the comments carefully and try to provide information
on all relevant titles.

#########################################################################
-->


## New Behavior

<!--
Please describe in a few words the intentions of your PR.
-->

This small patch add the support for the STORAGE_BACKEND and STORAGE_CONFIG variables
from the environment, already supported by netbox.

## Contrast to Current Behavior

<!--
Please describe in a few words how the new behavior is different
from the current behavior.
-->

Currently the variables are evaluated in the netbox code, but not read from the environement,
so they are ignored. I experimented this on both docker and kubernetes deployments.

## Discussion: Benefits and Drawbacks

<!--
Please make your case here:

- Why do you think this project and the community will benefit from your
  proposed change?
- What are the drawbacks of this change?
- Is it backwards-compatible?
- Anything else that you think is relevant to the discussion of this PR.

(No need to write a huge article here. Just a few sentences that give some
additional context about the motivations for the change.)
-->

This change will add an already supported feature, improving the scalability and statefulness
on kubernetes or docker deployments, allowing the removal of a local storage dependancy.

## Changes to the Wiki

<!--
If the README.md must be updated, please include the changes in the PR.
If the Wiki must be updated, please make a suggestion below.
-->

We could add a small comment or a link to the relevant section on the netbox docs, e.g:
[Netbox storage backend](https://netbox.readthedocs.io/en/stable/configuration/optional-settings/#storage_backend)

## Proposed Release Note Entry

<!--
Please provide a short summary of your PR that we can copy & paste
into the release notes.
-->

Add the support for the STORAGE_BACKEND and STORAGE_CONFIG variables
from the environment.
## Double Check

<!--
Please put an x into the brackets (like `[x]`) if you've completed that task.
-->

* [x] I have read the comments and followed the PR template.
* [x] I have explained my PR according to the information in the comments.
* [x] My PR targets the `develop` branch.

gervaso committed:
Add STORAGE_BACKEND support in configuration.py

This settings enable the support for STORAGE_BACKEND and STORAGE_CONFIG in the configuration.py file.

cimnine commented on configuration/configuration.py:
```diff
@@ -199,6 +199,17 @@ def read_secret(secret_name):
+# A dictionary of configuration parameters for the storage backend configured as STORAGE_BACKEND. The specific
+# parameters to be used here are specific to each backend; see the django-storages documentation for more detail.
+# If STORAGE_BACKEND is not defined, this setting will be ignored.
+STORAGE_CONFIG = os.environ.get('STORAGE_CONFIG', None)
```
> A dictionary of configuration parameters for the storage backend configured as STORAGE_BACKEND

This says the value of the STORAGE_CONFIG should be a dictionary. But ` os.environ.get` will return only strings. Do you have an example how to use it?

gervaso replied:
Searching for "STORAGE_CONFIG" [here](https://netbox.readthedocs.io/en/stable/release-notes/version-2.7/), shows an example, the expected syntax is like this one:
```
    STORAGE_CONFIG = {
    'AWS_ACCESS_KEY_ID': '<Key>',
    'AWS_SECRET_ACCESS_KEY': '<Secret>',
    'AWS_STORAGE_BUCKET_NAME': 'netbox',
    'AWS_S3_REGION_NAME': 'eu-west-1',
}
```

I added this config using extraEnvs for the helm values:
```yaml
extraEnvs:
  LOGIN_REQUIRED: true
  METRICS_ENABLED: true
  STORAGE_BACKEND: 'storages.backends.s3boto3.S3Boto3Storage'
  STORAGE_CONFIG: |
    {
    "AWS_S3_ENDPOINT_URL": "http://netbox-minio:9000",
    "AWS_S3_USE_SSL": False,
    "AWS_ACCESS_KEY_ID": "minio-access-key",
    "AWS_SECRET_ACCESS_KEY": "minio-secret-key",
    "AWS_STORAGE_BUCKET_NAME": "netbox-assets"
    }
```

The env vars values in the pod seemed correct, but I could not properly test them.

cimnine replied:
I don't mean to be rude, but did you successfully test the proposed feature before submitting it? Because in my understanding I don't see how it could work as it is.

In Python (`configuration.py` is just Python code), `os.environ.get` will return a [`string`](https://docs.python.org/3/library/string.html). But `STORAGE_CONFIG` is expected to be a [`Dictionary`](https://docs.python.org/3/tutorial/datastructures.html#dictionaries). Plus I suspect that `STORAGE_BACKEND` is expected to be a `class`, and not _the name of a class as `string`_.

Make no mistake, I could be wrong and just don't get it. That's why I ask you to either provide a working example (using the minio container seems a good idea to do that) or to amend your PR so that we get to a working configuration.

gervaso replied:
Don't worry, you were not rude at all and in fact you were right:  the STORAGE_CONFIG needs to be converted to a Dictionary.
The STORAGE_BACKEND appears to be correct though.
As I said before, I could test the vars values inside the k8s pods but I could not create an updated build incorporating the changes.

I have made some changes to the patch and I'll try to check everything on a new container image build and testing it with "plain" docker, asap.
Thanks for the feedback.

gervaso replied:
Hello again, I properly tested the feature and I confirm that it works as expected.
There's a minor issue: I had to add "boto3" to the requirements in the Dockerfile, but any backend supported by django-storages would require its own libraries.

Let me know if i could help any further.

gervaso committed:
add parsing of  dict-string

STORAGE_CONFIG must be converted from string to a Dictionary, only if STORAGE_BACKEND is defined.

cimnine commented:
Thanks for the PR. Unfortunately I've not had the time to look into PRs recently.

The changes you've provided look like it would work now. But as I explained in #310 I'm not a fan of using `eval` (or `literal_eval`). So as it stands I'm currently not comfortable merging the PR as it is. I (and everyone's invited to do so as well) will have to think about handling such dynamic configuration values in the future.

cimnine commented:
I would like to share one idea I just had and why I don't think it is _the_ solution either.

Take this env file for example:

```
# netbox.env
NETBOX_STORAGE_BACKEND = AWS
NETBOX_STORAGE_AWS_ACCESS_KEY_ID = xxx
NETBOX_STORAGE_AWS_SECRET_ACCESS_KEY = xxx
NETBOX_STORAGE_AWS_STORAGE_BUCKET_NAME = xxx
NETBOX_STORAGE_AWS_S3_REGION_NAME = xxx
NETBOX_STORAGE_AWS_S3_ENDPOINT_URL = https://minio
```

And in `configuration.py` something like:

```py
supported_storage_backends = {
  'AWS': 'storages.backends.s3boto3.S3Boto3Storage',
  'LIBCLOUD': '...',
  'AZURE': '...',
  # ...
}

STORAGE_BACKEND = supported_storage_backend.get(os.environ.get('NETBOX_STORAGE_BACKEND', None), None)

if STORAGE_BACKEND:
  STORAGE_CONFIG = {}
  for (key, value) in STORAGE_BACKEND.items():
    prefix = 'NETBOX_STORAGE_'
    if key.startswith(prefix  + STORAGE_BACKEND + '_'):
      # strip prefix "NETBOX_STORAGE_" to turn the key into e.g. "AWS_xxx"
      STORAGE_CONFIG[key[len(prefix):] = value
```

I think that this implementation would be an improvement over relying on `literal_eval`, because – while it's still generic – it is restricted to just the well-known storage prefixes. No piece of code can slip in and change things unexpectedly.

But I think it is still not good, because one would be forced to store secrets in environment variables – [something that should be avoided in any production environment](https://diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/) – without providing an option not to do so.

cimnine commented:
I've started a discussion issue here: https://github.com/netbox-community/netbox-docker/issues/318

cimnine commented:
Please let me know if #343 would be an acceptable solution or why not. Thanks!
